// @antv/l7-utils/es/lodash-adapter.js
import clamp from 'lodash/clamp';
// cloneDeep
import cloneDeep from 'lodash/cloneDeep';
import debounce from 'lodash/debounce';
// get
import get from 'lodash/get';
import isBoolean from 'lodash/isBoolean';
// isEqual
import isEqual from 'lodash/isEqual';
import isNil from 'lodash/isNil';
import isNumber from 'lodash/isNumber';
import isPlainObject from 'lodash/isPlainObject';
import isString from 'lodash/isString';
import isTypedArray from 'lodash/isTypedArray';
import merge from 'lodash/merge';
// mergewith
import mergeWith from 'lodash/mergeWith';
import pull from 'lodash/pull';
import throttle from 'lodash/throttle';
// uniq clamp
import uniq from 'lodash/uniq';
// upperFirst
import upperFirst from 'lodash/upperFirst';

// isFunction
import isFunction from 'lodash/isFunction';
// isObject
import isObject from 'lodash/isObject';
// isundefined
import isUndefined from 'lodash/isUndefined';

// camelCase
import camelCase from 'lodash/camelCase';
export var lodashUtil = {
  isNil: isNil,
  merge: merge,
  throttle: throttle,
  isString: isString,
  debounce: debounce,
  pull: pull,
  isTypedArray: isTypedArray,
  isPlainObject: isPlainObject,
  isNumber: isNumber,
  isBoolean: isBoolean,
  isEqual: isEqual,
  cloneDeep: cloneDeep,
  uniq: uniq,
  clamp: clamp,
  upperFirst: upperFirst,
  get: get,
  mergeWith: mergeWith,
  isFunction: isFunction,
  isObject: isObject,
  isUndefined: isUndefined,
  camelCase: camelCase
};
