// @antv/g6-core/lib/index.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  registerNode: true,
  registerCombo: true,
  registerEdge: true,
  registerBehavior: true,
  BaseGlobal: true,
  AbstractGraph: true,
  Shape: true,
  Arrow: true,
  Marker: true,
  Util: true,
  AbstractLayout: true,
  AbstractEvent: true,
  Node: true,
  Edge: true,
  Hull: true,
  Combo: true
};
Object.defineProperty(exports, "AbstractEvent", {
  enumerable: true,
  get: function get() {
    return _event.default;
  }
});
Object.defineProperty(exports, "AbstractGraph", {
  enumerable: true,
  get: function get() {
    return _graph.default;
  }
});
Object.defineProperty(exports, "AbstractLayout", {
  enumerable: true,
  get: function get() {
    return _layout.default;
  }
});
Object.defineProperty(exports, "Arrow", {
  enumerable: true,
  get: function get() {
    return _element.Arrow;
  }
});
exports.BaseGlobal = void 0;
Object.defineProperty(exports, "Combo", {
  enumerable: true,
  get: function get() {
    return _combo.default;
  }
});
Object.defineProperty(exports, "Edge", {
  enumerable: true,
  get: function get() {
    return _edge.default;
  }
});
Object.defineProperty(exports, "Hull", {
  enumerable: true,
  get: function get() {
    return _hull.default;
  }
});
Object.defineProperty(exports, "Marker", {
  enumerable: true,
  get: function get() {
    return _element.Marker;
  }
});
Object.defineProperty(exports, "Node", {
  enumerable: true,
  get: function get() {
    return _node.default;
  }
});
Object.defineProperty(exports, "Shape", {
  enumerable: true,
  get: function get() {
    return _element.default;
  }
});
Object.defineProperty(exports, "Util", {
  enumerable: true,
  get: function get() {
    return _util.default;
  }
});
exports.registerNode = exports.registerEdge = exports.registerCombo = exports.registerBehavior = exports.default = void 0;
var _behavior = _interopRequireDefault(require("./behavior"));
var _graph = _interopRequireDefault(require("./graph/graph"));
var _element = _interopRequireWildcard(require("./element"));
var _global = _interopRequireDefault(require("./global"));
var _util = _interopRequireDefault(require("./util"));
var _layout = _interopRequireDefault(require("./graph/controller/layout"));
var _event = _interopRequireDefault(require("./graph/controller/event"));
var _node = _interopRequireDefault(require("./item/node"));
var _edge = _interopRequireDefault(require("./item/edge"));
var _hull = _interopRequireDefault(require("./item/hull"));
var _combo = _interopRequireDefault(require("./item/combo"));
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
// 用于 PC 和 Mobile 端分别实现 layout 和 updateLayoutCfg 方法

var registerNode = exports.registerNode = _element.default.registerNode;
var registerEdge = exports.registerEdge = _element.default.registerEdge;
var registerCombo = exports.registerCombo = _element.default.registerCombo;
var registerBehavior = exports.registerBehavior = _behavior.default.registerBehavior;
var BaseGlobal = exports.BaseGlobal = _global.default;
var _default = exports.default = {
  version: _global.default.version,
  AbstractGraph: _graph.default,
  BaseGlobal: BaseGlobal,
  Util: _util.default,
  Shape: _element.default,
  Node: _node.default,
  Edge: _edge.default,
  Combo: _combo.default,
  Hull: _hull.default,
  registerNode: _element.default.registerNode,
  registerEdge: _element.default.registerEdge,
  registerCombo: _element.default.registerCombo,
  registerBehavior: _behavior.default.registerBehavior,
  Arrow: _element.Arrow,
  Marker: _element.Marker,
  AbstractLayout: _layout.default,
  AbstractEvent: _event.default
};
