// @antv/l7-renderer/es/device/DeviceAttribute.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
var DeviceAttribute = /*#__PURE__*/function () {
  function DeviceAttribute(device, options) {
    _classCallCheck(this, DeviceAttribute);
    var buffer = options.buffer,
      offset = options.offset,
      stride = options.stride,
      normalized = options.normalized,
      size = options.size,
      divisor = options.divisor,
      shaderLocation = options.shaderLocation;
    this.buffer = buffer;
    this.attribute = {
      shaderLocation: shaderLocation,
      buffer: buffer.get(),
      offset: offset || 0,
      stride: stride || 0,
      normalized: normalized || false,
      divisor: divisor || 0
    };
    if (size) {
      this.attribute.size = size;
    }
  }
  _createClass(DeviceAttribute, [{
    key: "get",
    value: function get() {
      return this.buffer;
    }
  }, {
    key: "updateBuffer",
    value: function updateBuffer(options) {
      this.buffer.subData(options);
    }
  }, {
    key: "destroy",
    value: function destroy() {
      this.buffer.destroy();
    }
  }]);
  return DeviceAttribute;
}();
export { DeviceAttribute as default };
